source/fundamentals/gridfs.txt

```diff
@@ -0,0 +1,395 @@
+of ``InputStream`` or write its data to a ``GridFSUploadStream``.
+
+For either upload process, you can specify configuration information such
+as file chunk size and other field/value pairs to store as metadata. Set
+this information on an instance of ``GridFSUploadOptions`` as shown in the
+following code snippet:
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start uploadOptions
+   :end-before: end uploadOptions
+
+See the :java-docs:`GridFSUploadOptions <apidocs/mongodb-driver-core/com/mongodb/client/gridfs/model/GridFSUploadOptions.html>`
+API documentation for more information.
+
+Upload a File Using an Input Stream
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section shows you how to upload a file to a GridFS bucket using an input
+stream. The code example below shows how you can use a ``FileInputStream`` to
+read data from a file in your filesystem and upload it to GridFS by performing
+the following operations:
+
+- Read from the filesystem using a ``FileInputStream``.
+- Set the chunk size using ``GridFSUploadOptions``.
+- Set a custom metadata field called ``type`` to the value "zip archive".
+- Upload a file called ``project.zip``, specifying the GridFS file name as "myProject.zip".
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start uploadFromInputStream
+   :end-before: end uploadFromInputStream
+
+This code example prints the file id of the uploaded file after it is
+successfully saved in GridFS.
+
+For more information, see the API documentation on
+:java-docs:`uploadFromStream() <apidocs/mongodb-driver-sync/com/mongodb/client/gridfs/GridFSBucket.html#uploadFromStream(java.lang.String,java.io.InputStream,com.mongodb.client.gridfs.model.GridFSUploadOptions)>`.
+
+Upload a File Using an Output Stream
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This section shows you how to upload a file to a GridFS bucket by writing to
+an output stream. The code example below shows how you can write to a
+``GridFSUploadStream`` to send data to GridFS by performing the following
+operations:
+
+- Read a file named "project.zip" from the filesystem into a byte array.
+- Set the chunk size using ``GridFSUploadOptions``.
+- Set a custom metadata field called ``type`` to the value "zip archive".
+- Write the bytes to a ``GridFSUploadStream``, assigning the file name
+  "myProject.zip". The stream reads data into a buffer until it reaches the
+  limit specified in the ``chunkSize`` setting, and inserts it as a new chunk
+  in the ``chunks`` collection.
+
+.. literalinclude:: /includes/fundamentals/code-snippets/GridFSOperations.java
+   :language: java
+   :dedent:
+   :start-after: start uploadFromOutputStream
+   :end-before: end uploadFromOutputStream
+
+This code example prints the file id of the uploaded file after it is
+successfully saved in GridFS.
+
```

Comment:
Suggestion add note about failure modes when uploading to GridFS and users having to clean up any orphaned chunks.

> Operation Failure
> If any of the above operations fail against the server, drivers MUST raise an error. If some inserts succeeded before the failed operation, these become orphaned chunks. Drivers MUST NOT attempt to clean up these orphaned chunks. The rationale is that whatever failure caused the orphan chunks will most likely also prevent cleaning up the orphaned chunks, and any attempts to clean up the orphaned chunks will simply cause long delays before reporting the original failure to the application.

See: https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#file-upload

Reply:
I'll add a paragraph on orphaned chunks. Would you recommend nesting a try block within the try-with-resources around the "write" method so that you can clean up using the GridFSUploadStream's abort() instance method in the catch block? Or that that they manually clean up orphaned chunks by directly querying and deleting them from the chunks collection?